File: easy/binary-tree-inorder-traversal.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def inorderTraversal(root):
    """
    If current does not have left child

        a. Add current’s value
    
        b. Go to the right, i.e., current = current.right

    Else

        a. In current's left subtree, make current the right child of the rightmost node

        b. Go to this left child, i.e., current = current.left
    """
    res = []
    curr = root
    stack = []
    if root is None:
        return res

    while curr is not None or len(stack) > 0:
        while curr is not None:
            stack.append(curr)
            curr = curr.left
        curr = stack.pop()
        res.append(curr.val)
        curr = curr.right

    return res


import unittest

class TestInorderTraversal(unittest.TestCase):

    def test_first_sample(self):
        root = TreeNode(val=1, left=None, right=TreeNode(val=2, left=TreeNode(val=3, left=None, right=None), right=None))
        self.assertEqual(inorderTraversal(root), [1,3,2])

    def test_second_sample(self):
        root = None
        self.assertEqual(inorderTraversal(root), [])

    def test_third_sample(self):
        root = TreeNode(val=1)
        self.assertEqual(inorderTraversal(root), [1])

    def test_fourth_sample(self):
        root = TreeNode(val=1, left=TreeNode(val=2, left=None, right=None), right=None)
        self.assertEqual(inorderTraversal(root), [2,1])

    def test_fourth_sample(self):
        root = TreeNode(val=2, left=TreeNode(val=3, left=TreeNode(val=1, left=None, right=None), right=None), right=None)
        self.assertEqual(inorderTraversal(root), [1,3,2])

if __name__ == '__main__':
    unittest.main()

File: easy/binary-tree-inorder-traversal.cpp
#include <memory>
#include <vector>
#include <stack>
#include <cassert>
#include <iostream>

class TreeNode {
public:
    int val;
    std::unique_ptr<TreeNode> left;
    std::unique_ptr<TreeNode> right;

    TreeNode(int val = 0, TreeNode* left = nullptr, TreeNode* right = nullptr)
        : val(val), left(left), right(right) {}
};

std::vector<int> inorderTraversal(TreeNode* root) {
    std::vector<int> res;
    std::stack<TreeNode*> stack;
    TreeNode* curr = root;

    while (curr != nullptr || !stack.empty()) {
        while (curr != nullptr) {
            stack.push(curr);
            curr = curr->left.get();
        }
        curr = stack.top();
        stack.pop();
        res.push_back(curr->val);
        curr = curr->right.get();
    }

    return res;
}

int main() {
    // Test 1
    {
        auto root = std::make_unique<TreeNode>(1, nullptr, new TreeNode(2, new TreeNode(3), nullptr));
        auto result = inorderTraversal(root.get());
        std::vector<int> expected = {1, 3, 2};
        assert(result == expected);
		std::cout << "Test 1 passes\n";
    }

    // Test 2
    {
        TreeNode* root = nullptr;
        auto result = inorderTraversal(root);
        std::vector<int> expected = {};
        assert(result == expected);
		std::cout << "Test 2 passes\n";
    }

    // Test 3
    {
        auto root = std::make_unique<TreeNode>(1);
        auto result = inorderTraversal(root.get());
        std::vector<int> expected = {1};
        assert(result == expected);
		std::cout << "Test 3 passes\n";
    }

    // Test 4
    {
        auto root = std::make_unique<TreeNode>(1, new TreeNode(2), nullptr);
        auto result = inorderTraversal(root.get());
        std::vector<int> expected = {2, 1};
        assert(result == expected);
		std::cout << "Test 4 passes\n";
    }

    // Test 5
    {
        auto root = std::make_unique<TreeNode>(2, new TreeNode(3, new TreeNode(1), nullptr), nullptr);
        auto result = inorderTraversal(root.get());
        std::vector<int> expected = {1, 3, 2};
        assert(result == expected);
		std::cout << "Test 5 passes\n";
    }

    return 0;
}
